src/sender_queue/mod.rs

```diff
@@ -237,6 +283,20 @@ where
                 .or_default()
                 .push(message);
         }
+        let mut removed_ids = Vec::new();
+        for (id, last_epoch) in &self.last_epochs {
+            if let Some(peer_epoch) = self.peer_epochs.get(id) {
+                if last_epoch <= peer_epoch {
+                    removed_ids.push(id.clone());
+                }
+            } else {
+                removed_ids.push(id.clone());
+            }
+        }
+        for id in removed_ids {
+            self.peer_epochs.remove(&id);
+            self.last_epochs.remove(&id);
+        }
```

Comment:
Is `defer_messages` the right place to check that? I see two triggers for removing an entry from `peer_epochs`:
* Either we handle a batch in which the peer is removed and their epoch is already at least that. Then we can remove it immediately. (In `update_epoch`.)
* Or we receive an `EpochStarted` message with the epoch in `last_epochs`, send them their remaining messages and then remove them. (In `process_new_epoch`.)

Reply:
Would you be happy with my current solution, that's waiting for `EpochStarted` from a first epoch after the last epoch of the old validator? Or would you like me to remove the old validator after all  locally produced messages for the last epoch have been sent to it?

Another possibility would be to remove in either of these two cases, whichever happens first.

Reply:
Implemented the latter.